### PR TITLE
View Transitions: do not animate same page navigation, but animate hash links to different pages

### PR DESCRIPTION
**Pull request by martrapp**
## Changes

- Formerly, links with a hash mark did not start view transitions. Now they will, if they lead to a different page.
- Formerly, links to the same page (without hashes) did start view transitions. Now they don't.  

## Testing

Tested with the examples from #7849 and #7882

## Docs

No doc updates required.
Fixes unexpected behavior

### CODE REVIEW

**Comment by martrapp** [comment]
There is still some minor issue:  `samePage("name", "name#")`  returns false, which makes links with `href="#"` to start a transition. (`href=""` and href=`#top` will not start transitions). Will add an additional fix for that to this PR.

**Comment by martrapp** [comment]
done.

File: packages/astro/components/ViewTransitions.astro [hunk 1]
```diff
@@ -220,6 +224,12 @@ const { fallback = 'animate' } = Astro.props as Props;
 		document.head.append(link);
 	}
 
+	function isSamePage(location, link) {
+		const linkURL = new URL(link.href, location.href);
+		linkURL.hash = location.hash;
+		return linkURL.href.replace(/#$/, '') === location.href.replace(/#$/, '');
```
Comment (matthewp) [inline packages/astro/components/ViewTransitions.astro:230]: Can you explain what this is doing? It's not obvious to me. Can we not just check if `location.pathname === link.pathname`?

Comment (martrapp) [reply packages/astro/components/ViewTransitions.astro:230]: Ha :smile: When I started, it seemed like the natural way to check if two URLs are "equal modulo the fragment ID". It got more complicated when I noticed a problem with hashes at the end.

But I probably mostly avoided worrying about other URL properties like searchParams.

I only have SSG experience with Astro.  
I suppose there are also SSR examples with fixed pathnames, where people expect different searchParams to show the transition to different pages?

Maybe we should aim for a simple, generic solution and give users an additional link property for manual control.

pathname === pathname is fine with me. I will fix it that way.

Maybe you already have ideas on how to override link behavior to force transitions for same pathnames. I'll leave that to a later change :-)

**Comment by matthewp** [comment]
@martrapp can you mark this as ready for review?

Also, does this close https://github.com/withastro/astro/issues/7849?

**Comment by martrapp** [comment]
> Also, does this close #7849?

yes, the example included in that issue now works as expected: `href=""` jumps to top of the page without triggering a view transition.

**Comment by ryansolid** [comment]
I think there are cases where you want to navigate to the same page and have it preserve.

1. Pagination using search parameters. This is where I noticed this fix messed up my Hackernews example. Things like `https://hackernews.com/top?page=2`. This also applies to search pages which will often refresh with different search parameters/filters.

2. Mutations. This one might be a different case but say posting a form to add a comment to a page while watching a video and not having the video unload. In this case there is no url difference but you just want to force it.

**Comment by matthewp** [comment]
Thanks @ryansolid, fixed in `astro@2.10.7`

**Pull request merged**